File: src/generators/LotsofLootLocationLootGenerator.ts
import { inject, injectable } from "tsyringe";

import { BaseClasses } from "@spt/models/enums/BaseClasses";
import { ConfigTypes } from "@spt/models/enums/ConfigTypes";

import { IContainerItem, LocationLootGenerator } from "@spt/generators/LocationLootGenerator";
import { ItemHelper } from "@spt/helpers/ItemHelper";
import { PresetHelper } from "@spt/helpers/PresetHelper";
import { ConfigServer } from "@spt/servers/ConfigServer";
import { DatabaseService } from "@spt/services/DatabaseService";
import { ItemFilterService } from "@spt/services/ItemFilterService";
import { HashUtil } from "@spt/utils/HashUtil";
import { MathUtil } from "@spt/utils/MathUtil";
import { ProbabilityObject, ProbabilityObjectArray, RandomUtil } from "@spt/utils/RandomUtil";
import type { ICloner } from "@spt/utils/cloners/ICloner";
import { LotsofLootItemHelper } from "../helpers/LotsofLootItemHelper";
import { LotsofLootConfig } from "../utils/LotsofLootConfig";
import { LotsofLootLogger } from "../utils/LotsofLootLogger";

import { IStaticAmmoDetails } from "@spt/models/eft/common/ILocation";
import { ISpawnpoint } from "@spt/models/eft/common/ILooseLoot";
import { IItem } from "@spt/models/eft/common/tables/IItem";
import { ITemplateItem } from "@spt/models/eft/common/tables/ITemplateItem";
import { ILocationConfig } from "@spt/models/spt/config/ILocationConfig";
import { ILootInlooseContainerLimitConfig } from "../models/ILotsofLootConfig";

@injectable()
export class LotsofLootLocationLootGenerator {
    private looseContainerItemFilterIndexCache: Record<string, string[]> = {};

    constructor(
        @inject("LocationLootGenerator") protected locationLootGenerator: LocationLootGenerator,
        @inject("ConfigServer") protected configServer: ConfigServer,
        @inject("DatabaseService") protected databaseService: DatabaseService,
        @inject("ItemFilterService") protected itemFilterService: ItemFilterService,
        @inject("ItemHelper") protected itemHelper: ItemHelper,
        @inject("PresetHelper") protected presetHelper: PresetHelper,
        @inject("PrimaryCloner") protected cloner: ICloner,
        @inject("HashUtil") protected hashUtil: HashUtil,
        @inject("RandomUtil") protected randomUtil: RandomUtil,
        @inject("MathUtil") protected mathUtil: MathUtil,
        @inject("LotsofLootItemHelper") protected lotsOfLootItemHelper: LotsofLootItemHelper,
        @inject("LotsofLootConfig") protected config: LotsofLootConfig,
        @inject("LotsofLootLogger") protected logger: LotsofLootLogger,
    ) {}

    public handleSpawningAlwaysSpawnSpawnpoint(spawnpoints: ISpawnpoint[], location: string): ISpawnpoint | null {
        switch (location) {
            case "tarkovstreets":
                //Make sure not to spawn it in the rusted key room
                const spawnsStreets = spawnpoints.filter((spawnpoint) => !spawnpoint.template.Id.includes("sh_ee_loot") && !spawnpoint.template.Id.includes("LotsOfLootRustedKeyRoom") && spawnpoint.template.Items.some((item) => item._tpl === "6582dbf0b8d7830efc45016f"));

                const spawnpointArrayStreets = new ProbabilityObjectArray<string, ISpawnpoint>(this.mathUtil, this.cloner);

                spawnsStreets.forEach((spawnpoint) => {
                    spawnpointArrayStreets.push(new ProbabilityObject(spawnpoint.template.Id, spawnpoint.probability, spawnpoint));
                });

                const drawStreets = spawnpointArrayStreets.draw(1, true)[0];

                return spawnpointArrayStreets.data(drawStreets);
            case "laboratory":
                // Hotfix for SPT not spawning this item on labs
                const spawnsLabs = spawnpoints.filter((spawnpoint) => spawnpoint.template.Id.includes("keywhite") && spawnpoint.template.Items.some((item) => item._tpl === "6711039f9e648049e50b3307"));

                const spawnpointArrayLabs = new ProbabilityObjectArray<string, ISpawnpoint>(this.mathUtil, this.cloner);

                spawnsLabs.forEach((spawnpoint) => {
                    spawnpointArrayLabs.push(new ProbabilityObject(spawnpoint.template.Id, spawnpoint.probability, spawnpoint));
                });

                const drawLabs = spawnpointArrayLabs.draw(1, true)[0];

                return spawnpointArrayLabs.data(drawLabs);
            default:
                return null;
        }
    }

    public createStaticLootItem(tpl: string, staticAmmoDist: Record<string, IStaticAmmoDetails[]>, parentId: string = undefined): IContainerItem {
        const itemTemplate = this.getItemTemplate(tpl);
        if (!itemTemplate) {
            return {
                items: [],
                width: 0,
                height: 0,
            };
        }

        let width = itemTemplate._props.Width;
        let height = itemTemplate._props.Height;
        let items: IItem[] = [
            {
                _id: this.hashUtil.generate(),
                _tpl: tpl,
                parentId: parentId,
            },
        ];

        // Handle different item types
        if (this.itemHelper.isOfBaseclass(tpl, BaseClasses.WEAPON)) {
            const rootItem = items[0];
            items = this.handleWeaponItem(items, tpl, staticAmmoDist);

            // Set proper width and height on weapon
            const itemSize = this.itemHelper.getItemSize(items, rootItem._id);
            width = itemSize.width;
            height = itemSize.height;
        } else if (this.itemHelper.isOfBaseclass(tpl, BaseClasses.MONEY) || this.itemHelper.isOfBaseclass(tpl, BaseClasses.AMMO)) {
            const stackCount = this.randomUtil.getInt(itemTemplate._props.StackMinRandom, itemTemplate._props.StackMaxRandom);
            items[0].upd = { StackObjectsCount: stackCount };
        } else if (this.itemHelper.isOfBaseclass(tpl, BaseClasses.AMMO_BOX)) {
            this.itemHelper.addCartridgesToAmmoBox(items, itemTemplate);
        } else if (this.itemHelper.isOfBaseclass(tpl, BaseClasses.MAGAZINE)) {
            this.handleMagazineItem(items, itemTemplate, staticAmmoDist);
        } else if (this.itemHelper.isOfBaseclass(tpl, BaseClasses.SIMPLE_CONTAINER) && tpl !== "5c093e3486f77430cb02e593") {
            this.handleContainerOrBackpackItem(items, staticAmmoDist, this.config.getConfig().lootinLooseContainer.lootInContainerModifier);
        } else if (this.itemHelper.isOfBaseclass(tpl, BaseClasses.BACKPACK)) {
            this.handleContainerOrBackpackItem(items, staticAmmoDist, this.config.getConfig().lootinLooseContainer.lootInBackpackModifier);
        } else if (this.itemHelper.armorItemCanHoldMods(tpl)) {
            const defaultPreset = this.presetHelper.getDefaultPreset(tpl);
            if (defaultPreset) {
                const presetAndMods: IItem[] = this.itemHelper.replaceIDs(defaultPreset._items);
                this.itemHelper.remapRootItemId(presetAndMods);
                presetAndMods[0].parentId = items[0].parentId;
                items = presetAndMods;
            } else if (itemTemplate._props.Slots?.length > 0) {
                items = this.itemHelper.addChildSlotItems(items, itemTemplate, this.configServer.getConfig<ILocationConfig>(ConfigTypes.LOCATION).equipmentLootSettings.modSpawnChancePercent);
            }
        }

        return {
            items: items,
            width: width,
            height: height,
        };
    }

    private getItemTemplate(tpl: string): ITemplateItem | null {
        const gotItem = this.itemHelper.getItem(tpl);
        return gotItem[0] ? gotItem[1] : null;
    }

    // Cobbled together mostly from SPT's way of generating these.
    private handleWeaponItem(items: IItem[], tpl: string, staticAmmoDist: Record<string, IStaticAmmoDetails[]>): IItem[] {
        const rootItem = items[0];

        // Get the original weapon preset
        const weaponPreset = this.cloner.clone(this.presetHelper.getDefaultPreset(tpl));
        if (weaponPreset?._items) {
            const itemWithChildren = this.itemHelper.reparentItemAndChildren(weaponPreset._items[0], weaponPreset._items);

            if (itemWithChildren?.length > 0) {
                items = this.itemHelper.reparentItemAndChildren(rootItem, itemWithChildren);
            }
        }

        const magazine = items.find((x) => x.slotId === "mod_magazine");

        if (magazine && this.randomUtil.getChance100(this.configServer.getConfig<ILocationConfig>(ConfigTypes.LOCATION).magazineLootHasAmmoChancePercent)) {
            // Get required templates
            const magTemplate = this.itemHelper.getItem(magazine._tpl)[1];
            const weaponTemplate = this.itemHelper.getItem(tpl)[1];
            const defaultWeapon = this.itemHelper.getItem(rootItem._tpl)[1];

            // Fill the magazine with cartridges
            const magazineWithCartridges: IItem[] = [magazine];
            this.itemHelper.fillMagazineWithRandomCartridge(magazineWithCartridges, magTemplate, staticAmmoDist, weaponTemplate._props.ammoCaliber, this.configServer.getConfig<ILocationConfig>(ConfigTypes.LOCATION).minFillStaticMagazinePercent / 100, defaultWeapon._props.defAmmo, defaultWeapon);

            // Replace the original magazine with the filled version
            items.splice(items.indexOf(magazine), 1, ...magazineWithCartridges);
        }

        return items;
    }

    private handleMagazineItem(items: IItem[], itemTemplate: ITemplateItem, staticAmmoDist: Record<string, IStaticAmmoDetails[]>): void {
        if (this.randomUtil.getChance100(this.configServer.getConfig<ILocationConfig>(ConfigTypes.LOCATION).magazineLootHasAmmoChancePercent)) {
            const magazineWithCartridges: IItem[] = [items[0]];
            this.itemHelper.fillMagazineWithRandomCartridge(magazineWithCartridges, itemTemplate, staticAmmoDist, null, this.configServer.getConfig<ILocationConfig>(ConfigTypes.LOCATION).minFillStaticMagazinePercent / 100);
            items.splice(0, 1, ...magazineWithCartridges);
        }
    }

    private handleContainerOrBackpackItem(items: IItem[], staticAmmoDist: Record<string, IStaticAmmoDetails[]>, modifier: number): void {
        const containerLoot = this.createLootInLooseContainer(items[0]._tpl, items[0]._id, staticAmmoDist, modifier);
        containerLoot.forEach((containerItem) => items.push(containerItem));
    }

    public createLootInLooseContainer(tpl: string, id: string, staticAmmoDist: Record<string, IStaticAmmoDetails[]>, modifier = 0.5): IItem[] {
        if (modifier === 0) {
            return [];
        }

        const tables = this.databaseService.getTables();
        const items = tables.templates.items;
        const item = items[tpl];

        // Ensure filters exist for item
        if (!item._props.Grids[0]._props.filters[0]) {
            this.logger.debug(`${item._name} doesn't have a filter, setting default filter!`);
            item._props.Grids[0]._props.filters = [
                {
                    Filter: ["54009119af1c881c07000029"],
                    ExcludedFilter: [],
                },
            ];
        }

        // Clone filters
        let whitelist = [...item._props.Grids[0]._props.filters[0].Filter];
        let blacklist = new Set(item._props?.Grids[0]._props.filters[0]?.ExcludedFilter || []);

        const amount = this.randomUtil.getInt(1, item._props.Grids[0]._props.cellsH * item._props.Grids[0]._props.cellsV * modifier);
        let fill = 0;

        // Use cache for whitelist if available
        if (this.looseContainerItemFilterIndexCache[tpl]) {
            whitelist = this.looseContainerItemFilterIndexCache[tpl];
        } else {
            this.logger.debug(`${tpl} is new, generating whitelist`);

            // Expand whitelist and blacklist with children items
            const newWhiteList = this.expandItemsWithChildrenItemIds(whitelist, items);
            const newBlackList = this.expandItemsWithChildrenItemIds(Array.from(blacklist), items);

            whitelist = newWhiteList;
            blacklist = new Set(newBlackList);

            if (this.config.getConfig().lootinLooseContainer?.blacklist[tpl]) {
                blacklist = new Set([...blacklist, ...this.config.getConfig().lootinLooseContainer.blacklist[tpl]]);
            }

            // Remove blacklist items from whitelist
            whitelist = whitelist.filter((item) => !blacklist.has(item));

            // Remove invalid items and built-in inserts from whitelist
            // biome-ignore format: Readability
            whitelist = whitelist.filter((itemTpl) => !this.itemHelper.isOfBaseclass(itemTpl, BaseClasses.BUILT_IN_INSERTS) && 
                !this.itemFilterService.isItemBlacklisted(itemTpl) && 
                !this.itemFilterService.isItemRewardBlacklisted(itemTpl) &&
                this.itemHelper.isValidItem(itemTpl) &&
                items[itemTpl]?._props.Prefab.path !== ""
            );

            // Cache the result for later reuse
            this.looseContainerItemFilterIndexCache[tpl] = whitelist;
        }

        // Return here as we cant put loose loot into containers that have no whitelisted items.
        if (whitelist.length === 0) {
            this.logger.warning(`${tpl} whitelist is empty`);
            return [];
        }

        let itemArray = new ProbabilityObjectArray<string, number>(this.mathUtil, this.cloner);
        whitelist.forEach((itemId) => {
            let itemWeight = 1;

            const price = tables.templates.prices[itemId];
            if (price) {
                itemWeight = Math.round(1000 / Math.pow(price, 1 / 3));
            } else if (itemId === "5449016a4bdc2d6f028b456f") {
                // Roubles
                itemWeight = 500;
            } else if (["5696686a4bdc2da3298b456a", "569668774bdc2da2298b4568"].includes(itemId)) {
                // Dollars, Euros
                itemWeight = 100;
            }

            itemArray.push(new ProbabilityObject(itemId, itemWeight));
        });

        // Generate loot items
        const generatedItems: IItem[] = [];
        const limits: ILootInlooseContainerLimitConfig = this.config.getConfig().lootinLooseContainer.spawnLimits[tpl] ?? null;
        let drawnKeys = 0;
        let drawnKeycards = 0;
        while (fill <= amount) {
            // Since we modify these with limits, check each loop if they are empty and if so break out from the while loop
            if (itemArray.length === 0 || whitelist.length === 0) {
                break;
            }

            // Handle if we should draw an item from the ProbabilityObjectArray (Weighted) or from the whitelist
            const drawnItemTpl = this.config.getConfig().general.itemWeights ? itemArray.draw(1, true)[0] : whitelist[this.randomUtil.getInt(0, whitelist.length - 1)];

            if (limits) {
                if (limits.keys != null) {
                    if (this.itemHelper.isOfBaseclass(drawnItemTpl, BaseClasses.KEY_MECHANICAL)) {
                        if (drawnKeys >= limits.keys) {
                            if (this.config.getConfig().general.itemWeights) {
                                itemArray = itemArray.filter((itemTpl) => !this.itemHelper.isOfBaseclass(itemTpl.key, BaseClasses.KEY_MECHANICAL));
                            } else {
                                whitelist = whitelist.filter((itemTpl) => !this.itemHelper.isOfBaseclass(itemTpl, BaseClasses.KEY_MECHANICAL));
                            }

                            continue;
                        }

                        drawnKeys++;
                    }
                }

                if (limits.keycards != null) {
                    if (this.itemHelper.isOfBaseclass(drawnItemTpl, BaseClasses.KEYCARD)) {
                        if (drawnKeycards >= limits.keycards) {
                            if (this.config.getConfig().general.itemWeights) {
                                itemArray = itemArray.filter((itemTpl) => !this.itemHelper.isOfBaseclass(itemTpl.key, BaseClasses.KEYCARD));
                            } else {
                                whitelist = whitelist.filter((itemTpl) => !this.itemHelper.isOfBaseclass(itemTpl, BaseClasses.KEYCARD));
                            }

                            continue;
                        }

                        drawnKeycards++;
                    }
                }
            }

            const lootItem = this.createStaticLootItem(drawnItemTpl, staticAmmoDist, id);
            lootItem.items[0].slotId = "main";
            fill += lootItem.height * lootItem.width;

            if (fill > amount) {
                break;
            }

            generatedItems.push(...lootItem.items);
        }

        return generatedItems;
    }

    private expandItemsWithChildrenItemIds(itemsToExpand: string[], items: Record<string, any>): string[] {
        const expandedItems: string[] = [];

        itemsToExpand.forEach((content) => {
            const childItems = this.lotsOfLootItemHelper.findAndReturnChildrenItemIdsByItems(items, content);
            expandedItems.push(...childItems);
        });

        return expandedItems;
    }
}

File: LotsofLoot/Generators/LotsofLootLocationLootGenerator.cs
using LotsofLoot.Helpers;
using LotsofLoot.Models.Config;
using LotsofLoot.Services;
using LotsofLoot.Utilities;
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.Extensions;
using SPTarkov.Server.Core.Generators;
using SPTarkov.Server.Core.Helpers;
using SPTarkov.Server.Core.Models.Common;
using SPTarkov.Server.Core.Models.Eft.Common;
using SPTarkov.Server.Core.Models.Eft.Common.Tables;
using SPTarkov.Server.Core.Models.Enums;
using SPTarkov.Server.Core.Models.Spt.Config;
using SPTarkov.Server.Core.Models.Spt.Inventory;
using SPTarkov.Server.Core.Models.Spt.Server;
using SPTarkov.Server.Core.Servers;
using SPTarkov.Server.Core.Services;
using SPTarkov.Server.Core.Utils;
using SPTarkov.Server.Core.Utils.Cloners;
using SPTarkov.Server.Core.Utils.Collections;

namespace LotsofLoot.Generators
{
    [Injectable(InjectionType.Singleton)]
    public class LotsofLootLocationLootGenerator(ItemHelper itemHelper, PresetHelper presetHelper, ItemFilterService itemFilterService,
        HashUtil hashUtil, RandomUtil randomUtil, ICloner cloner, LotsofLootItemHelper LotsofLootItemHelper,
        LotsOfLootLogger logger, ConfigService config, DatabaseService databaseService, ConfigServer configServer)
    {
        private readonly Dictionary<string, List<MongoId>> _itemFilterIndexCache = [];
        private readonly HashSet<MongoId> _foreignCurrencies =
        [
            // Dollars
            "5696686a4bdc2da3298b456a",
            // Euros
            "569668774bdc2da2298b4568"
         ];

        /// <summary>
        /// This method closely mirrors that of SPT
        /// The only difference being the bypass for loot overlay and using Lots of Loot's createStaticLootItem
        /// </summary>
        public List<SpawnpointTemplate> GenerateDynamicLoot(LooseLoot dynamicLootDist, Dictionary<string, IEnumerable<StaticAmmoDetails>> staticAmmoDist, string locationName)
        {
            return [];
        }

        //Todo: Fully needs implementing
        public Spawnpoint? HandleSpawningAlwaysSpawnSpawnpoint(List<Spawnpoint> spawnpoints, string location)
        {
            return null;
        }

        public ContainerItem CreateStaticLootItem(string chosenTpl, Dictionary<string, IEnumerable<StaticAmmoDetails>> staticAmmoDist,
            string? parentId = null)
        {
            TemplateItem? itemTemplate = GetItemTemplate(chosenTpl);

            if (itemTemplate is null || itemTemplate.Properties is null)
            {
                logger.Warning($"{chosenTpl} has no template?");

                return new ContainerItem
                {
                    Items = [],
                    Width = 0,
                    Height = 0,
                };
            }

            int? width = itemTemplate.Properties.Width;
            int? height = itemTemplate.Properties.Height;

            List<Item> items =
            [
                new() {
                    Id = new MongoId(),
                    Template = chosenTpl,
                    ParentId = parentId,
                }
            ];

            // Handle different item types
            if (itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.WEAPON))
            {
                Item rootItem = items[0];
                items = HandleWeaponItem(items, chosenTpl, staticAmmoDist);

                // Set proper width and height on weapon
                ItemSize itemSize = itemHelper.GetItemSize(items, rootItem.Id);
                width = itemSize.Width;
                height = itemSize.Height;
            }
            else if (itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.MONEY) || itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.AMMO))
            {
                int stackCount = randomUtil.GetInt((int)itemTemplate.Properties.StackMinRandom, (int)itemTemplate.Properties.StackMaxRandom);
                items[0].Upd = new Upd
                {
                    StackObjectsCount = stackCount 
                };
            }
            else if (itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.AMMO_BOX))
            {
                itemHelper.AddCartridgesToAmmoBox(items, itemTemplate);
            }
            else if (itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.MAGAZINE))
            {
                HandleMagazineItem(items, itemTemplate, staticAmmoDist);
            }
            else if (itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.SIMPLE_CONTAINER) && chosenTpl != "5c093e3486f77430cb02e593")
            {
                HandleContainerOrBackpackItem(items, staticAmmoDist, config.LotsOfLootConfig.LootinLooseContainer.LootInContainerModifier);
            }
            else if (itemHelper.IsOfBaseclass(chosenTpl, BaseClasses.BACKPACK))
            {
                HandleContainerOrBackpackItem(items, staticAmmoDist, config.LotsOfLootConfig.LootinLooseContainer.LootInBackpackModifier);
            }
            else if (itemHelper.ArmorItemCanHoldMods(chosenTpl))
            {
                Preset? defaultPreset = presetHelper.GetDefaultPreset(chosenTpl);

                if (defaultPreset != null)
                {
                    List<Item> presetAndMods = defaultPreset.Items.ReplaceIDs().ToList();
                    presetAndMods.RemapRootItemId();
                    presetAndMods[0].ParentId = items[0].ParentId;
                    items = presetAndMods;
                }
                else if (itemTemplate.Properties.Slots?.Count() > 0)
                {
                    items = itemHelper.AddChildSlotItems(items, itemTemplate, configServer.GetConfig<LocationConfig>().EquipmentLootSettings.ModSpawnChancePercent);
                }
            }

            return new ContainerItem
            {
                Items = items,
                Width = width,
                Height = height,
            };
        }

        private TemplateItem? GetItemTemplate(string itemTpl)
        {
            KeyValuePair<bool, TemplateItem?> item = itemHelper.GetItem(itemTpl);

            if (item.Key)
            {
                return item.Value;
            }
            else
            {
                return null;
            }
        }

        private List<Item> HandleWeaponItem(List<Item> items, string tpl, Dictionary<string, IEnumerable<StaticAmmoDetails>> staticAmmoDist)
        {
            Item rootItem = items[0];

            // Get the original weapon preset
            Preset? weaponPreset = cloner.Clone(presetHelper.GetDefaultPreset(tpl));
            if (weaponPreset?.Items != null)
            {
                List<Item> itemWithChildren = itemHelper.ReparentItemAndChildren(weaponPreset.Items[0], weaponPreset.Items);

                if (itemWithChildren != null && itemWithChildren.Count > 0)
                {
                    items = itemHelper.ReparentItemAndChildren(rootItem, itemWithChildren);
                }
            }

            Item? magazine = items.Find(x => x.SlotId == "mod_magazine");

            if (magazine != null && randomUtil.GetChance100(configServer.GetConfig<LocationConfig>().MagazineLootHasAmmoChancePercent))
            {
                // Get required templates
                TemplateItem? magTemplate = itemHelper.GetItem(magazine.Template).Value;
                TemplateItem? weaponTemplate = itemHelper.GetItem(tpl).Value;
                TemplateItem? defaultWeapon = itemHelper.GetItem(rootItem.Template).Value;

                // Fill the magazine with cartridges
                List<Item> magazineWithCartridges = [magazine];

                itemHelper.FillMagazineWithRandomCartridge(
                    magazineWithCartridges,
                    magTemplate,
                    staticAmmoDist,
                    weaponTemplate.Properties.AmmoCaliber,
                    configServer.GetConfig<LocationConfig>().MinFillStaticMagazinePercent / 100.0,
                    defaultWeapon.Properties.DefAmmo,
                    defaultWeapon);

                // Replace the original magazine with the filled version
                var magIndex = items.IndexOf(magazine);
                items.RemoveAt(magIndex);
                items.InsertRange(magIndex, magazineWithCartridges);
            }

            return items;
        }

        private void HandleMagazineItem(List<Item> items, TemplateItem itemTemplate, Dictionary<string, IEnumerable<StaticAmmoDetails>> staticAmmoDist)
        {
            if (!randomUtil.GetChance100(configServer.GetConfig<LocationConfig>().MagazineLootHasAmmoChancePercent))
            {
                return;
            }

            List<Item> magazineWithCartridges = [items[0]];

            itemHelper.FillMagazineWithRandomCartridge(
                magazineWithCartridges,
                itemTemplate,
                staticAmmoDist,
                null,
                configServer.GetConfig<LocationConfig>().MinFillStaticMagazinePercent / 100.0);

            items.RemoveAt(0);
            items.InsertRange(0, magazineWithCartridges);
        }

        private void HandleContainerOrBackpackItem(List<Item> items, Dictionary<string, IEnumerable<StaticAmmoDetails>> staticAmmoDist, double modifier)
        {
            List<Item> containerLoot = CreateLootInLooseContainer(items[0].Template, items[0].Id, staticAmmoDist, modifier);

            foreach (var containerItem in containerLoot)
            {
                items.Add(containerItem);
            }
        }

        public List<Item> CreateLootInLooseContainer(string tpl, MongoId id, Dictionary<string, IEnumerable<StaticAmmoDetails>> staticAmmoDist, double modifier = 0.5)
        {
            if (modifier == 0)
            {
                return [];
            }

            DatabaseTables tables = databaseService.GetTables();
            Dictionary<MongoId, TemplateItem>? items = databaseService.GetTables()?.Templates?.Items;

            if (!items.TryGetValue(tpl, out var item))
            {
                logger.Warning($"Template {tpl} not found in database.");
                return [];
            }

            // Ensure filters exist for item
            var firstGrid = item.Properties.Grids.FirstOrDefault();
            if (firstGrid == null)
                return [];

            var firstFilter = firstGrid.Props?.Filters?.FirstOrDefault();

            if (firstFilter == null || !firstGrid.Props.Filters.Any())
            {
                logger.Debug($"{item.Name} doesn't have a filter, setting default filter!");
                firstGrid.Props.Filters =
                [
                    new GridFilter
                    {
                        Filter = ["54009119af1c881c07000029" ],
                        ExcludedFilter = []
                    }
                ];
                firstFilter = firstGrid.Props.Filters.First(); // reset after assigning
            }

            // Clone filters
            List<MongoId> whitelist = [.. firstFilter.Filter];
            HashSet<MongoId> blacklist = [.. firstFilter.ExcludedFilter ?? []];

            int maxCells = (int)(firstGrid.Props.CellsH * firstGrid.Props.CellsV);
            int amount = randomUtil.GetInt(1, (int)(maxCells * modifier));

            // Use cache for whitelist if available, if not available, generate new cache
            if (!_itemFilterIndexCache.TryGetValue(tpl, out var cachedWhiteList))
            {
                logger.Debug($"{tpl} is new, generating whitelist");

                // Expand items with children
                whitelist = ExpandItemsWithChildItemIds(whitelist, items);
                List<MongoId> expandedBlacklist = ExpandItemsWithChildItemIds(blacklist.ToList(), items);
                blacklist.UnionWith(expandedBlacklist);

                // Add config blacklist
                if (config.LotsOfLootConfig.LootinLooseContainer.Blacklist.TryGetValue(tpl, out var configBlacklist))
                {
                    blacklist.UnionWith(configBlacklist);
                }

                // Filter whitelist - single pass instead of multiple Where calls
                whitelist = whitelist.Where(itemTpl =>
                    !blacklist.Contains(itemTpl) &&
                    !itemHelper.IsOfBaseclass(itemTpl, BaseClasses.BUILT_IN_INSERTS) &&
                    !itemFilterService.IsItemBlacklisted(itemTpl) &&
                    !itemFilterService.IsItemRewardBlacklisted(itemTpl) &&
                    itemHelper.IsValidItem(itemTpl) &&
                    items.TryGetValue(itemTpl, out var itm) &&
                    !string.IsNullOrEmpty(itm.Properties.Prefab?.Path)
                ).ToList();

                // Cache whitelist
                _itemFilterIndexCache[tpl] = whitelist;
            }
            else
            {
                whitelist = cachedWhiteList;
            }

            if (whitelist.Count == 0)
            {
                logger.Warning($"{tpl} whitelist is empty");
                return [];
            }

            // Build probability array with improved weight calculation
            ProbabilityObjectArray<MongoId, int?> itemArray = new(cloner);
            Dictionary<MongoId, double>? prices = tables.Templates?.Prices;

            foreach (var itemId in whitelist)
            {
                int itemWeight = 1;

                if (itemId == "5449016a4bdc2d6f028b456f")
                {
                    itemWeight = 500;
                }
                else if (_foreignCurrencies.Contains(itemId))
                {
                    itemWeight = 100;
                }
                else if (prices.TryGetValue(itemId, out var price))
                {
                    itemWeight = (int)Math.Round(1000 / Math.Pow(price, 1.0 / 3.0));
                }

                itemArray.Add(new ProbabilityObject<MongoId, int?>(itemId, itemWeight, null));
            }

            // Generate loot items
            List<Item> generatedItems = [];
            LootInLooseContainerSpawnLimit? limits = config.LotsOfLootConfig.LootinLooseContainer.SpawnLimits.TryGetValue(tpl, out LootInLooseContainerSpawnLimit? lim) ? lim : null;

            int fill = 0;
            int drawnKeys = 0;
            int drawnKeycards = 0;

            while (fill <= amount)
            {
                // Since we modify these with limits, check each loop if they are empty and if so break out from the while loop
                if (itemArray.Count == 0 || whitelist.Count == 0)
                {
                    break;
                }

                // Handle if we should draw an item from the ProbabilityObjectArray (Weighted) or from the whitelist
                string drawnItemTpl = config.LotsOfLootConfig.General.ItemWeights
                    ? itemArray.DrawAndRemove(1).FirstOrDefault()
                    : whitelist[randomUtil.GetInt(0, whitelist.Count - 1)];

                // Check limits if they exist
                if (limits != null)
                {
                    if (limits.Keys.HasValue && itemHelper.IsOfBaseclass(drawnItemTpl, BaseClasses.KEY_MECHANICAL))
                    {
                        if (drawnKeys >= limits.Keys.Value)
                        {   
                            if (config.LotsOfLootConfig.General.ItemWeights)
                            {
                                itemArray = itemArray.Filter(i => !itemHelper.IsOfBaseclass(i.Key, BaseClasses.KEY_MECHANICAL));
                            }
                            else
                            {
                                whitelist = whitelist.Where(i => !itemHelper.IsOfBaseclass(i, BaseClasses.KEY_MECHANICAL)).ToList();
                            }
                        }
                        else
                        {
                            drawnKeys++;
                        }
                    }

                    if (limits.Keycards.HasValue && itemHelper.IsOfBaseclass(drawnItemTpl, BaseClasses.KEYCARD))
                    {
                        if (drawnKeycards >= limits.Keycards.Value)
                        {
                            if (config.LotsOfLootConfig.General.ItemWeights)
                            {
                                itemArray = itemArray.Filter(i => !itemHelper.IsOfBaseclass(i.Key, BaseClasses.KEYCARD));
                            }
                            else
                            {
                                whitelist = whitelist.Where(i => !itemHelper.IsOfBaseclass(i, BaseClasses.KEYCARD)).ToList();
                            }
                        }
                        else
                        {
                            drawnKeycards++;
                        }
                    }
                }

                ContainerItem lootItem = CreateStaticLootItem(drawnItemTpl, staticAmmoDist, id);
                lootItem.Items.First().SlotId = "main";
                fill += (int)(lootItem.Height * lootItem.Width);

                if (fill > amount)
                {
                    break;
                }

                generatedItems.AddRange(lootItem.Items);
            }

            return generatedItems;
        }

        private List<MongoId> ExpandItemsWithChildItemIds(List<MongoId> itemsToExpand, Dictionary<MongoId, TemplateItem> items)
        {
            List<MongoId> expandedItems = [];

            foreach (MongoId content in itemsToExpand)
            {
                List<MongoId> childItems = LotsofLootItemHelper.FindAndReturnChildItemIdsByItems(items, content);
                expandedItems.AddRange(childItems);
            }

            return expandedItems;
        }
    }
}
